### ISSUE TITLE
Enhance distributedLoad

### ISSUE THREAD

**Issue by Binyang2014** [Alluxio/alluxio#13485]
**Summary**
Currently the max parallel jobs for distributedLoad is controlled by `ACTIVE_JOB_COUNT_OPTION`. For folder which contains many small files, such as imageNet, this number is not easy to define.

If the number is too large, many jobs will failed (due to the requests limitation of UFS). If the number is too small, the load speed will be impacted. 

Here we propose a policy to adjust this number automatically. 

**Policy**

The basic idea is like TCP congestion control.  https://en.wikipedia.org/wiki/TCP_congestion_control
If some job failed at a period, we set the active_job_num to half of original value and set the thredhold to this number.
If all jobs are succeed at a period,  and current number under the threshold we double the active_job_num,  else we linear increase the active_job_num.

Required init parameters:  `ACTIVE_JOB_MAX_NUM`, `ACTIVE_JOB_INIT_NUM`, `ACTIVE_JOB_INCREASE_PACE ` , `RUNING_JOB_MAX_NUM`

```
ACTIVE_JOB_INIT_NUM = init_active_job_num
ACTIVE_JOB_MAX_NUM = max_active_job_num
ACTIVE_JOB_INCREASE_PACE  = increase_pace
RUNING_JOB_MAX_NUM = running_job_max_num

active_job_num: current number of active jobs, the initial value is ACTIVE_JOB_INIT_NUM 
activate_job_threshold, the threshold for active job
active_jobs: a queue for all active jobs
```
```
The main load porecss:

for file_name in load_file_list
   create job with corresponded file name
   if active_jobs.size() < active_job_num
     add job into active_jobs
   else
     wait a while
   end
end
drain active_jobs, waiting for all job finished.
```

```
submission process
using multi-thread, for each thread:
get job from active_jobs
submit job to job master
```

```
adjust active job number process
run in another thread
while !finished
  wake_up every second
  completed_job_num
  failed_job_num
  for job in active_jobs
    if job in completed status
      completed_job_num++
    else if job in failed status
      failed_job++
    end
  end
  if failed_job != 0
    active_job_num = max(active_job_num/2, init_active_job_num)
    activate_job_threshold = active_job_num 
  else if active_job_num - completed_job_num > running_job_max_num
      // too many jobs in running status, don't adjust the nunber
  else if active_job_num > activate_job_threshold
     active_job_num  = active_job_num  + ACTIVE_JOB_INCREASE_PACE   // linear increase
  else
     active_job_num   = min(active_job_num  * 2, ACTIVE_JOB_MAX_NUM) // double the number
   end
end
```

**Comment by Binyang2014** [comment]
@apc999 @LuQQiu Please help to review this design

**Comment by Binyang2014** [comment]
Tested in our bed, this policy works well for small files. For 1 millions files, the failed tasks is around 100

**Comment by LuQQiu** [comment]
@Binyang2014 Some comments for the design
1. Can we make it optional, and recommend to be used when loading small files?
2. I think our current distributedLoad command has retry mechanism for the failed tasks, can your design maintain the retry logics especially when load failed because we launching too many active jobs?
3. Can we maintain the backward compatibility that the older version `bin/alluxio distributedLoad` can work with this new version distributedLoad command just like before?

Other than these three comments, the design LGTM. It's awesome to have this automation functionality

**Comment by yuzhu** [comment]
@Binyang2014 awesome idea. one question i had was whether you have any thoughts on fairness. It does not have to be addressed now but it would be nice to give it some thought. 
The current design allows for a single distributedLoad job to overwhelm the entire ufs. How do we adapt if there are multiple distributedLoad jobs loading from different ufs at the same time? i think the maxJobNumber might need to be per mount point? or per ufs?

have you done any tests or simulations where two distributed load jobs are happening at the same time?

**Comment by Binyang2014** [comment]
For Lu's comments:
1. Sure, we can make it optional. And I think it works for both small file and large file. For large file, since we set the `running_job_max_num`, the number of job will be limited. Maye we can give some recommend init parameters for small file scenario and large file scenario.
2. For current design, the retry logic is kept. We only treat the task failed if it reaches the max retry number. Otherwise, we will treat it as an active job.
3. Maybe we can add a option for current `distributedLoad` command. Such as `bin/alluxio distributedLoad -dynamicAdjustOptoin {activeJobInitNum: xx, activeJobMaxNum: xx}` to achieve backward compatibility.

**Comment by Binyang2014** [comment]
For @yuzhu's comments:
Actually, I don't think much about fairness, I just put some thoughts here.
1.  The fairness issue may need be handled by external services. For many scenarios, distributedLoad happens before real workload scheduled to the compute cluster. And the scheduler has the whole view of the workload (which workload will be scheduled first and which job has the high priority ). Then job scheduler need to decide how to allocation resource for distributedLoad. For alluxio, we need to expose some interface for scheduler to allocate resources (such as maxJobNum...). And report the current status to let external scheduler adjust its' decision.
2. For which resource need to be exposed to scheduler, it's a bit complex. `maxJobNum` may be one dimension (compute dimension), and the `consumed bandwidth`  maybe another. (We need to make sure the prefetch job not consume too much bandwidth which may affect current running jobs)

**Comment by LuQQiu** [comment]
The design LGTM, @yuzhu PTAL thanks!

**Comment by yuzhu** [comment]
@Binyang2014 the scenario I imagined is as follows. it will  limit our bandwidth unnecessarily. 
Imagine two concurrent distributedLoad jobs,  each accessing a different ufs. 
If the first one reaches network capacity, and starts to error, it would prevent the second job from increasing the maxJobNum even if it is no where near the limit

**Comment by Binyang2014** [comment]
Hi @yuzhu. for each distributedLoad job. The bandwidth consumption can be caped by parameter `ACTIVE_JOB_MAX_NUM`. And from my point, this parameter should be select by admin or job scheduler based on current cluster status to avoid occupy too much bandwidth.

For alluxio, it may need to report some metrics to let scheduler/admin do this decision.

And for future enhance, the  `ACTIVE_JOB_MAX_NUM` parameter may not enough to cap the bandwidth, we need to investigate and add more parameters to control this. (just like we can limit the cpu/memory/disk usage for container)

**Comment by yuzhu** [comment]
yeah, I think in general multi tenancy/ QOS concerns in job service requires further investigation. the current improvement LGTM, as we are no worse than before. @Binyang2014
